kpr: Enable/fix Cilium socket based load balancing in different environments

aditighag opened the pull request:
This PR aims to revisit some of the assumptions made around cgroup hierarchies in Cilium in order to enable socket-lb in different environments.

## Context 

Cilium attaches `BPF_CGROUP_*` type programs to provide socket based load-balancing. The default cgroup root in the agent is set to a custom location (`/var/run/cilium/cgroupv2`), where the agent tries to mount cgroup filesystem. The cgroup root is then passed to `init.sh` in order to attach the `BPF_CGROUP_*` programs at relevant hook points. While we have some extended logic in place to accommodate environments like `kind`, the overall logic breaks in certain scenarios. While the following list is not an exhaustive list, it helps in identifying general patterns.

## Scenarios where current logic breaks

- Virtualized cgroup root in the cgroup namespace mode (#15137)
If container runtimes are run with cgroup v2, Cilium agent pod would be deployed in a separate cgroup namespace. For example, Docker container runtime with cgroupv2 support switched to [private](https://docs.docker.com/config/containers/runmetrics/#running-docker-on-cgroup-v2) cgroup namespace mode as the default. Due to [cgroup namespaces](https://man7.org/linux/man-pages/man7/cgroup_namespaces.7.html), the cgroup fs mounted by the Cilium pod points to a virtualized cgroup hierarchy instead of the host cgroup root. As a result, BPF programs are attached to the nested cgroup root, and socket-lb isn't effective for other pods. 
**Resolution:** 
Mount cgroup fs on the host from init containers. We need to specify cgroup as the [enterable namespace](https://man7.org/linux/man-pages/man1/nsenter.1.html) in the nsenter command. Cilium agent will auto mount cgroup2 fs on the underlying host if not already mounted. This requires mounting host's `/proc` inside an init container temporarily. As an alternative, users can disable auto mount, and specify a mount point on the host where cgroup2 fs is already mounted. Cgroup2 fs mount point is platform dependent. Hence, we introduce a new helm option for the host cgroup2 fs mount point. See this note in [cgroup man page](https://man7.org/linux/man-pages/man7/cgroups.7.html)  -
Note that on many modern systems, systemd(1) automatically mounts
the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
process.

- Commit 866969e3 hard-coded `kubelet` string which may not work in some platforms. Depending on the value of kubelet config `--cgroup-root`, this string may or may not be present. Moreover, the logic to get cgroup root is specific to `kind` environments only so that it won't take effect for minikube clusters.
**Resolution:** 
The kind cgroup root detection logic can be removed as kind nodes and cilium agent pod are deployed in the same cgroup ns (if they don't, the above init container fix should be sufficient).


- Deploying a kind cluster alongside "other" BPF cgroup programs (#16078)
I ran into this issue while deploying a kind cluster on the dev VM. Cilium pod inside the kind cluster fails to come up with this warning message `failed to attach program`, and is stuck in a "CrashLoopback" state. We can have `cilium/ebpf` loader print better error messages/hints in such cases (I'll file an issue). 
I traced the error return code `255` (aka operation not permitted) from to this check [1] in the kernel source that disallows attach in the presence of programs (no override/multi) in the parent cgroup. After I removed the BPF programs attached in the dev VM cgroup root, I was able to create a kind cluster with socket-lb successfully.
[1] https://elixir.bootlin.com/linux/latest/source/kernel/bpf/cgroup.c#L457
**Resolution:** 
We document this case along with potential steps to resolve the issue. 


## Testing
Tested the changes on kind, GKE and bare metal k8s cluster by verifying that BPF programs are correctly attached, and socket-lb works as expected.

Deferred to follow-ups - 
 - Move BPF fs to an init container in the cilium ds
 - Remove the agent code that mounts cgroup v2 fs if not mounted : This is done in order to not break installation with cilium cli. Cilium CLI crafts its own DS, and doesn't use the OSS DS. Once we've released a new version of the CLI (https://github.com/cilium/cilium-cli/issues/256) with similar fixes as this PR, we can safely remove this.
 - CI/dev env #16528
 - Extend bugtool to collect output of attached cgroup programs

Fixes: #16078 
Fixes: #15769 
Fixes: 866969e3e43931c5d45b9b1a4e601f5bdf7895d4
Fixes: https://github.com/cilium/cilium/issues/15137
(Reported-by : @kkourt) 

**Release note**
```release-note
Fixes connectivity issues when kube-proxy replacement is enabled, caused by 
ineffective socket based load balancing (aka host reachable services) in the private 
cgroup namespace mode of container runtimes (e.g., docker cgroupv2 configuration). 
```

aditighag committed:
pkg/cgroup: Fix a typo

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
cilium-daemonset: Fix ineffective socket-lb caused by incorrect cgroup2 fs mount

If container runtimes are run with cgroup v2, Cilium agent pod would be deployed in a
separate cgroup namespace. For example, Docker container runtime with cgroupv2 support
switched to private cgroup namespace mode as the default [1]. Due to cgroup namespaces [2],
the cgroup fs mounted by the Cilium pod points to a virtualized cgroup hierarchy instead of
the host cgroup root. As a result, BPF programs are attached to the nested cgroup root, and socket-lb
isn't effective for other pods.

Fix:

Mount cgroup2 fs from the host so that BPF programs are attached at the host cgroup root.
A new init container is added to the Cilium Daemonset that mounts cgroup2 fs on the host.
The `/proc/1/ns/` directory on the host is required to be mounted so that cgroup and mount
namespaces are enabled as enterable namespaces while running the `nsenter` command.
Additionally, cgroup2 fs can be attached to different paths so let's mount it on the host
at a cilium-specific custom location. Cilium can thus have control over the location
(e.g., create the directory if it doesn't exist). This also helps in effectively identifying
if a cgroup2 mount already exists at the custom location.

[1] https://docs.docker.com/config/containers/runmetrics/#running-docker-on-cgroup-v2
[2] https://man7.org/linux/man-pages/man7/cgroup_namespaces.7.html

Reported-By: Kornilios Kourtis <kornilios@isovalent.com>
Fixes: #15137

Signed-off-by: Aditi Ghag <aditi@cilium.io>

kkourt reviewed (commented):
Thanks for the patch!

The patch at its current state does not seem to work for me [1]. it's probably because cilium tries to mount in `/var/run/cilium` and it uses this directory (which is not the global one). Passing a `--cgroup-root` also does not seem to work because then the detection will not run. 

I'm confident that we can get the patch to work, but I would like to propose to  leave the detection logic outside of the cilium agent, and just pass a proper argument with --cgroup-root. We can even pass two options, one for the root and one for where to attach the bpf programs, if we care about this distinction. We already had a special case for Kind, now we are going to add another one, and I think we should keep this complexity out of the agent.

[1] This is the logs:

```
level=info msg="Mounted cgroupv2 filesystem at /var/run/cilium/cgroupv2" subsys=cgroups
level=warning msg="Failed to determine cgroup v2 hierarchy for Kind node. Socket-based LB (--enable-host-reachable-services) will not work." error="cannot find \"kubelet\" in cgroup v2 path: \"/var/run/cilium/cgroupv2\"" subsys=cgroups
```

aditighag committed:
Revert "cgroups: Determine cgroup v2 hierarchy root for Kind"

This reverts commit e9ce8306400bf416087046b8d5b013b23ebdcb3e.

This logic is no longer needed as we mount cgroup v2 filesystem
from the underlying kubernetes node. This will enable cilium to
correctly attach BPF programs at every `kind` node's cgroup root.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Document failure scenario for kind deployment

Deploying a kind cluster in an environment where Cilium is already
running (for example, in the Cilium development VM) can lead to Cilium
pods crashing. This can also happen if there are other BPF cgroup programs
attached to the parent ``cgroup`` hierarchy of the kind container nodes.

Relevant Linux kernel code reference - https://elixir.bootlin.com/linux/latest/source/kernel/bpf/cgroup.c#L457.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Add troubleshooting steps to the kube-proxy free guide

Document the requirement that Cilium agent needs to be able
to attach BPF cgroup programs at the host cgroup root, in order
for socket-based load balancing (aka host-reachable services)
to be effective for other pods and host processes.

More details in the PR - https://github.com/cilium/cilium/pull/16259

Signed-off-by: Aditi Ghag <aditi@cilium.io>

qmonnet commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -370,6 +372,16 @@ spec:
 {{- end }}
       hostNetwork: true
       initContainers:
+        {{/* Required to mount cgroup v2 filesystem on the underlying Kubernetes node.*/}}
```
If using a template comment, please [add spaces](https://helm.sh/docs/chart_best_practices/templates/#formatting-templates) after `{{` and before `}}`.
```suggestion
      {{ /* Required to mount cgroup v2 filesystem on the underlying Kubernetes node */ }}
```

But should this just be a [YAML comment instead](https://helm.sh/docs/chart_best_practices/templates/#comments-yaml-comments-vs-template-comments)? It doesn't document something specific to the template?

qmonnet commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -427,6 +439,10 @@ spec:
 {{- /* Required for wait-bpf-mount to work */}}
           mountPropagation: HostToContainer
 {{- end }}
+{{/*Required to mount cgroup filesystem from the host to cilium agent pod*/}}
```
Same here.

qmonnet commented on Documentation/gettingstarted/kind.rst:
```diff
@@ -81,6 +81,24 @@ and no longer routing api-server requests to the current ``kind-control-plane``
+by following the `bpftool documentation <https://manpages.ubuntu.com/manpages/focal/man8/bpftool-cgroup.8.html>`_.
+For more information, see the `Pull Request <https://github.com/cilium/cilium/pull/16259>`__.
+
+.. parsed-literal::
```
```suggestion
::
```

Nothing to be parsed in this literal block.

qmonnet commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1205,6 +1205,43 @@ As per `k8s Service <https://kubernetes.io/docs/concepts/services-networking/ser
+effective leading to connectivity issues.
+For more information, ensure that you have the fix `Pull Request <https://github.com/cilium/cilium/pull/16259>`__.
+
+.. parsed-literal::
```
```suggestion
.. code-block:: shell-session
```

qmonnet reviewed (changes requested):
PR looks good to me, but I have a few nits on formatting.

qmonnet reviewed (approved):
All good from my side, thank you!

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -458,6 +479,16 @@ spec:
+      - hostPath:
+          path: /proc
+          type: DirectoryOrCreate
+        name: host-proc
```
We shouldn't mount host's `/proc` into Cilium's DaemonSet.

brb replied:
A bit similar as above, this is missing _why_.

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -370,6 +374,19 @@ spec:
+          - mountPath: /hostproc
+            name: host-proc
+        securityContext:
+          privileged: true
```
This is a really nasty hack.

brb replied:
IMHO saying that this is _a really nasty hack_ doesn't provide much helpful information. Mind elaborating instead?

aditighag replied:
>Bidirectional mount propagation as proposed in the weekly meeting?

@aanm Can you explain how the propagation mount will solve the cgroup namespace problem? Also, cgroup v2 fs is not guaranteed to be mounted on the host at `/sys/fs/cgroup` path. 
Example on GKE - 
```
$ mount | grep cgroup2
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime)
```

kkourt replied:
One thing to note is that if we know _where_ the cgroupv2 host fs is mounted in the host, the mount propagation allows us to expose the host cgroupfs inside the agent's container without having to expose /proc [1]. Couldn't we have a configurable helm variable for this location (default /sys/fs/cgroup/) that can be set to a proper value for each platform that deviates (e.g., GKE).

E.g., something like:

```
- hostPath:
       path: {{ .Values.hostCgroupFS }} # /sys/fs/cgroup by default
        type: DirectoryOrCreate
   name: cilium-cgroup
```

and:
```
- mountPath: /var/run/cilium/cgroupv2
   name: cilium-cgroup
```


[1] I've just tested this. I think the way it works is by  doing a bind mount inside the root fs of the container.

aditighag replied:
I think this is may be a fair middle ground. Having said that, I may have a preference for mounting cgroupfs at a cilium specified custom path - 
As far as possible, we should try to keep the internal details as transparent to the end user as possible. It can be error prone to have too many knobs. See this note in the [cgroup manpage](https://man7.org/linux/man-pages/man7/cgroups.7.html) - 

>Note that on many modern systems, systemd(1) automatically mounts
       the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
       process.

So the cgroupfs mount location may differ across cgroup v1/v2, systemd or not, etc. The default` /sys/fs/cgroup` path for cgroupfs 2 doesn't necessarily cover a large number of environments.

aanm replied:
> IMHO saying that this is _a really nasty hack_ doesn't provide much helpful information. Mind elaborating instead?

@brb As mentioned in the review description, bidirectional mounts offered by Kubernetes should help having the behavior that we want without having the mount the entire `/proc`, increasing attack surface, into Cilium's container.

Sure there are environments where the cgroup directory is not the default one but we can hard-code the value of `--cgroup-root` for the environments that we know of. We already have dedicated options for cloud providers so I don't see how much difference it would be having dedicated value for the `--cgroup-root` depending on the cloud provider. We can even mount `/sys/fs`, instead of `/sys/fs/bpf`, inside the cilium-agent container and we could detect if the cgroup is mounted in ` /sys/fs/cgroup/unified` or in ` /sys/fs/cgroup`.

For the more specific environments where users might change the cgroup directory intentionally then we just need to add documentation regarding this in the kube-proxy GSG.

aditighag replied:
>without having the mount the entire /proc, increasing attack surface, into Cilium's container.

We are only mounting `/proc` inside the init container that's short-lived (not inside the agent container). Cilium CLI already [mounts](https://github.com/cilium/cilium-cli/blob/master/install/install.go#L685) `/proc` in order to mount `/sys/fs/bpf`.

aanm replied:
> > without having the mount the entire /proc, increasing attack surface, into Cilium's container.
> 
> We are only mounting `/proc` inside the init container that's short-lived (not inside the agent container). Cilium CLI already [mounts](https://github.com/cilium/cilium-cli/blob/master/install/install.go#L685) `/proc` in order to mount `/sys/fs/bpf`.

Cilium CLI is still in beta stage, the reason it is mounting proc was to make it easier to deploy Cilium across all environments but that does not mean it's the safest way of mounting bpf fs. We should replace Cilium cli with bidirectional mounting instead of mounting `/proc` in Cilium repo.

brb replied:
To reduce the attack surface, wouldn't it be enough to mount `/proc/1/ns/cgroup` and `/proc/1/ns/mnt` instead of `/proc`?

aditighag replied:
Yeah, we could do selective `/proc/*` mounts.

aditighag replied:
`/proc/1/ns/cgrou`p and `/proc/1/ns/cgroup` are symlinks. I've made the change to mount `/proc/1/ns` instead.

aanm reviewed (changes requested):
Did you check an alternative to the hack using [Bidirectional](https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation) mount propagation as proposed in the weekly meeting?

aditighag committed:
defaults: Update default cgroup root

`/var/run` is a symlink to `/run` on most platforms, and may
not always be present. Also, this is consistent with the
`DefaultMapRootFallback` currently configured in the agent.

Example -

$ sudo mount -t cgroup2 none /var/run/cilium/cgroupv2
$ mount | grep cgroup
none on /run/cilium/cgroupv2 type cgroup2 (rw,relatime)

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1201,6 +1201,44 @@ As per `k8s Service <https://kubernetes.io/docs/concepts/services-networking/ser
 This can be allowed by setting ``bpf.lbExternalClusterIP=true``.
 
+
+Troubleshooting
```
@brb I think we also need to add a note about incompatibility between cgroup v1 and v2 to the installation step in this guide?
Edit: Checked with Daniel offline, this won't be an issue when this PR is merged since we'll be attaching programs to the cgroup root. We already have a note in `kind` gsg.

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -370,6 +374,21 @@ spec:
+          - name: CGROUP_ROOT
+            value: {{ .Values.cgroupRoot }}
+        command: ['nsenter', '--cgroup=/hostpid1ns/cgroup', '--mount=/hostpid1ns/mnt', "--",
+                  'sh', '-c',  'mount | grep "$CGROUP_ROOT type cgroup2" || { echo "Mounting cgroup filesystem..."; mount -t cgroup2 none $CGROUP_ROOT; }']
```
nit, can we have this in a multiline?

aanm commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1643,3 +1643,10 @@ clustermesh:
+
+# -- Enable auto mount of cgroup v2 filesystem
+cgroupAutoMount:
+  enabled: true
```
Why not have the options like:
```
cgroup:
  autoMount:
    enabled: true
  root: /run/cilium/cgroupv2
```

aanm commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1643,3 +1643,10 @@ clustermesh:
   enabled: false
+
+# -- Configure cgroup root where cgroup v2 filesystem is mounted
+cgroupRoot: /run/cilium/cgroupv2
```
I don't understand this one. With the current version of the `master` branch, the cgroupv2 are already available in this directory. Why do we need to specify it as an helm flag?

aditighag replied:
This helm value corresponds to the Cilium agent config option (`cgroup-root`) that allows users to specify a different path than the default.

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -426,6 +445,12 @@ spec:
           mountPropagation: HostToContainer
+{{- end }}
+        # Required to mount cgroup filesystem from the host to cilium agent pod
+{{- if .Values.cgroupAutoMount }}
```
The comment should be inside the if statement

brb commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -457,6 +486,24 @@ spec:
+{{- if .Values.cgroupAutoMount.enabled }}
+      # To mount cgroup filesystem on the host
+      - hostPath:
+          path: /proc/1/ns
```
Could we limit the mount scope even more? I.e. to mount only `/proc/1/ns/cgroup` and `/proc/1/ns/mnt`?

aditighag replied:
Those are just symlinks.

brb replied:
From namespaces(7):
>        Bind mounting (see mount(2)) one of the files in this directory
>       to somewhere else in the filesystem keeps the corresponding
>       namespace of the process specified by pid alive even if all
>       processes currently in the namespace terminate.
To my understanding, mounting only those files should work.

aditighag replied:
@brb I'm not sure that's supposed to work. I'm unable to mount `/proc/1/ns/mnt`. It may be because of the circular dependency. This might be related- https://stackoverflow.com/questions/34783391/why-i-couldnt-use-mount-bind-proc-pid-ns-mnt-to-another-file-in-ubuntu

`  Warning  Failed     20s                kubelet            Error: failed to start container "mount-cgroup": Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"rootfs_linux.go:58: mounting \\\"/proc/1/ns/mnt\\\" to rootfs \\\"/var/lib/docker/overlay2/7b5659481002373db731ac03d0563a9cad37be0027d77bf755a8332a32c080e4/merged\\\" at \\\"/var/lib/docker/overlay2/7b5659481002373db731ac03d0563a9cad37be0027d77bf755a8332a32c080e4/merged/hostpid1mntns\\\" caused \\\"invalid argument\\\"\"": unknown`

aanm replied:
@aditighag try changing the type from `DirectoryOrCreate` to `Directory`

aditighag replied:
@aanm I'm not sure which line you were pointing to. Anyway I meant to fix the type for `/proc/1/ns mount`. The DS shouldn't try to create that directory. 😅 ✅

aanm replied:
>Could we limit the mount scope even more? I.e. to mount only /proc/1/ns/cgroup and /proc/1/ns/mnt?

To limiting the scope as Martynas mentioned above I would suggest trying to set `type: File` as it was previously set with `DirectoryOrCreate`.

kkourt commented on Documentation/helm-values.rst:
```diff
@@ -90,6 +90,18 @@
+   * - cgroup
+     - Configure cgroup related configuration
+     - object
+     - ``{"autoMount":{"enabled":true},"root":"/run/cilium/cgroupv2"}``
```
Is this a different option than `cgroup.root`? If yes, do we need both of them?

kkourt commented on Documentation/helm-values.rst:
```diff
@@ -90,6 +90,18 @@
+     - bool
+     - ``true``
+   * - cgroup.root
+     - Configure cgroup root where cgroup v2 filesystem is mounted
```
The above line seems a bit confusing to me. How about: "Location where the cgroup v2 filesystem is mounted inside the cilium agent container"?

kkourt commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -276,6 +276,10 @@ spec:
           name: bpf-maps
 {{- end }}
+{{- if not (contains "/run/cilium/cgroupv2" .Values.cgroup.root) }}
+        - mountPath: {{ .Values.cgroup.root }}
```
If I'm reading this correctly, `cgroup.root` is used to refer to locations (mountpoints) that, depending on its value, are in the host or in the agent's container. I think that's confusing.

aditighag replied:
(I'm replying to @kkourt 's comment here so that we can have the discussion in a single thread.)

Chatted with Kornilios offline. Looks like there was some confusion around the PR.

If auto-mount is enabled, then cgroup2 fs will be mounted at `cgroupRoot` on both the host and the agent pod. Else,  cgroup2 fs mounted at `cgroupHostRoot` is mounted inside the agent pod at `cgroupRoot`. I can clarify this a bit further.

>one configuration variable that points to the location of the cgroup dir on the host, to be used when autoMount is disabled.

This is the `cgroupHostRoot` that the PR has introduced. 

>one configuration variable for the cgroup root inside the container, which is what directory that the agent will use.

If this config only points to the directory inside the agent pod, why do we need to expose that as an helm option to users?

Alternatively, we can drop the additional option `cgroupHostRoot`. We hand-waived over some of these details in the datapath meeting.

option 1
- cgroup.autoMount : Enable/disable
- cgroup.root : If `autoMount` is enabled, then agent will mount cgroup2 fs at that path. Else, it'll mount that directory on the host inside the agent pod. In both the cases, we allow users to override the path. However, we'll need to have different defaults in auto-mount enable (`/run/cilium/cgroupv2`) and disable (`/sys/fs/cgroup`) cases. 

option 2
- cgroup.autoMount : Enable/disable
If `autoMount` is enabled, then agent will mount cgroup2 fs at a custom (hard-coded) location `/run/cilium/cgroupv2`.  Users won't be able to override this. 
- cgroup.root : This will be used when users disable auto-mount, and agent will then mount host's `cgroup.root` inside the agent pod. The default will be `/sys/fs/cgroup`.

I think option 2 is simpler.

joestringer replied:
> If this config only points to the directory inside the agent pod, why do we need to expose that as an helm option to users?

:heavy_plus_sign: 

Option 2 sounds simpler and more consistent to me. If `cgroup.autoMount=false`, would we mount `cgroup.root` to `/run/cilium/cgroupv2` inside the container?

kkourt replied:
@aditighag, thanks for the clarification! Indeed I misunderstood the PR.

Both options are fine by me. The only comment I would make is that `cgroup.root` seems somewhat similar to the agent option `-cgroup-root` so I would suggest we use a different name to avoid confusion. `cgroup.hostRoot` maybe?

aditighag replied:
> If cgroup.autoMount=false, would we mount cgroup.root to /run/cilium/cgroupv2 inside the container?

This is exactly why I introduced a new option (`cgroupHostRoot`) in the current commits. If we go with option 2, we'll need to have `/run/cilium/cgroupv2` hard-coded in the agent DS, or have the same mount path as the host. Unless folks have some other ideas?

joestringer replied:
If you don't specify it anywhere, then the default agent value is `""` which it then replaces with `DefaultCgroupRoot = "/var/run/cilium/cgroupv2"`. That seems fine to me.

aditighag replied:
I was referring to the agent daemonset yaml file changes when auto-mount is enabled. 
I'm getting conflicting comments on the PR. 😕 

Where do we want to mount cgroup2 fs inside the cilium agent pod when `auto.mount` is - 
- enabled
- disabled 

If it's going to be `/run/cilium/cgroupv2` in both the cases, then we can have the path hard-coded, and remove the `cgroup-root` agent configmap option. If we want it to be `/run/cilium/cgroupv2` in the 1st case, but same as the host cgroup root for the latter case, then don't we have the same confusion if we rename the current option to `cgroupHostRoot`? 

IMO, we are overly complicating this. Let's just call it `cgroupRoot`, and have it better documented.

joestringer replied:
One thing that made it more complicated is it's the same actual directory in the cilium container:

```
# ls -lh /var/run
lrwxrwxrwx 1 root root 4 Apr 16 05:11 /var/run -> /run
```

I think we should stick with `/var/run/cilium/cgroupv2` so we're consistent with the defaults in the daemon, in case for whatever reason these directories diverge at some point in the future.

Then in the DaemonSet, if `cgroup.automount` is enabled, we do the nsenter trick to mount it to this path. If `cgroup.automount` is disabled, then we specify a cgroup hostPath under `volumes` as the configured argument `cgroup.root` or whatever we call it. Only in the latter case we need a `volumeMount` to mount the path to `/var/run/cilium/cgroupv2` (which will then be automatically used by the daemon defaults without having to specify it anywhere in configmap or DS commandline args).

I'm fine with `cgroup.root` or `cgroup.hostRoot` in the helm config, either way I agree that documenting that it's about the host path is the important piece. I don't think many users care how it looks inside the container.

joestringer replied:
Oh, I think I glossed over some of the detail in the automount side which makes the above comment miss some of the nuance.

For automount I think the goal is to mount at a known Cilium path on the host, then mount that to the consistent internal path. I _think_ if `cgroup.automount=true` it's fine to automount the cgroupfs to either `/var/run/cilium/cgroupv2` or `cgroup.root` on the host, I don't know if makes a big difference.  Then in that case we also need to specify the `volumes` / `volumeMount`, based on wherever we decide the cgroup FS should be automounted. Maybe I'm finally coming around to why this is complicated :sweat_smile:

joestringer replied:
^^ Probably makes sense to have some default for `cgroup.root` and always use that as the host path, either that we try to mount cgroupfs there in the case that `cgroup.automount=true`, or we use that (expected pre-mounted) path to mount into the container in the case that `cgroup.automount=false`. Then we can decide whatever the default for `cgroup.root` should be.

kkourt reviewed (commented):
Thanks Aditi!

Maybe I'm missing something, but the helm options seem confusing to me.

IMHO, we should have two different configuration options:
* one configuration variable that points to the location of the cgroup dir on the host, to be used  when autoMount is disabled.
* one configuration variable for the cgroup root inside the container, which is what directory that the agent will use.

kkourt commented on Documentation/helm-values.rst:
```diff
@@ -90,6 +90,18 @@
+     - bool
+     - ``true``
+   * - cgroup.root
+     - Configure cgroup root where cgroup v2 filesystem is mounted
```
```suggestion
     - Configure cgroup root where cgroup v2 filesystem is mounted on the host (see also: `cgroup.autoMount`)
```

kkourt reviewed (changes requested):
As I mentioned elsewhere, I think the helm option `cgroup.root` is too close to the agent option `-cgroup-root`. @aditighag I believe you suggested `cgroup.hostRoot` somewhere. Can we go with that?

Thanks.

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -458,6 +487,16 @@ spec:
+          path: /proc/1/ns
+          type: DirectoryOrCreate
+        name: host-proc-ns
+      # To keep state between restarts / upgrades for cgroup2 filesystem
+      - hostPath:
+          path: {{ .Values.cgroup.hostRoot}}
+          type: DirectoryOrCreate
+        name: cilium-cgroup
```
Is this really necessary? I thought the cgroup2 state would be kept even if it was not mounted in the host. I might be wrong but that was my impression.

aditighag replied:
>cgroup2 state would be kept even if it was not mounted in the host

Not sure I understand. Where would the state be kept?

aanm replied:
oh this is the hostRoot ok, nevermind.

aanm commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1643,3 +1643,8 @@ clustermesh:
+
+# -- Configure cgroup related configuration
+cgroup:
+  # -- Configure cgroup root where cgroup2 filesystem is mounted on the host (see also: `cgroup.autoMount`)
+  hostRoot: /run/cilium/cgroupv2
```
shouldn't this be by default `/sys/fs/cgroup`? The comment is a little confusing. Looking at the helm chart (install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml line 281) it looks that this is used to define the path where cgroup-root is mounted inside the container.

edit since I was reviewing the PR per commit basis: It also seems that with `cgroup.autoMount=false` Cilium will not mount the cgroup v2 correctly under `/sys/fs/cgroup` as there is no reference of it anywhere.

aditighag replied:
See this thread - https://github.com/cilium/cilium/pull/16259#discussion_r656486823.  😄 

@kkourt suggested to rename the option. This makes sense since it is user-facing. Users shouldn't care where we mount the fs inside the agent pod.

>It also seems that with cgroup.autoMount=false Cilium will not mount the cgroup v2 correctly under /sys/fs/cgroup

See this comment - 

```
# If users disable `autoMount`, it's expected that users have mounted
# cgroup2 filesystem at the specified `cgroup.hostRoot` 
```

aanm replied:
> See this thread - [#16259 (comment)](https://github.com/cilium/cilium/pull/16259#discussion_r656486823). 
> 
> @kkourt suggested to rename the option. This makes sense since it is user-facing. Users shouldn't care where we mount the fs inside the agent pod.
> 
> > It also seems that with cgroup.autoMount=false Cilium will not mount the cgroup v2 correctly under /sys/fs/cgroup
> 
> See this comment -
> 
> ```
> # If users disable `autoMount`, it's expected that users have mounted
> # cgroup2 filesystem at the specified `cgroup.hostRoot` 
> ```

But why isn't the default `/sys/fs/cgroup` and `/sys/fs/cgroup/unified` for gke? This would work as fine regardless if `autoMount` is true or false. However if we have `/run/cilium/cgroupv2` set as default, this will not work by default for users running with `autoMount` as false, which we can avoid it by setting the default to `/sys/fs/cgroup` for `/sys/fs/cgroup/unified` for gke.

joestringer replied:
Defaulting to mounting the cgroupfs to `/sys/fs/cgroup` on the host sounds fine to me, I doubt that there's much problem there - it's the canonical path so it's not like we'd get some other filesystem mounted there. If it's already mounted, we can detect that and avoid mounting it. If it's not yet mounted, then I don't think anyone would complain if we mounted it. Every mount of the cgroupv2 FS points to the same underlying cgroupfs implementation so there's nothing special about us mounting it there vs any other software mounting it there, so from that reasoning I don't think there's a compatibility issue or anything.

aanm replied:
@aditighag, while I was discussing this topic with Joe we are a little bit confused about one thing: We thought the whole point of `autoMount=true` was to auto detect the cgroup root mount point, so why are we using `/run/cilium/cgroupv2` in `autoMount=true`? Or is a misunderstanding from our part?

aditighag replied:
@joestringer To be clear, I haven't changed the default path. You explained the reasoning in the sig-datapath meeting why we chose to mount cgroup2 fs at a custom location. 

@aanm It's auto mount, where the init-container will automatically **mount** cgroup2 fs.

aanm replied:
>@aanm It's auto mount, where the daemonset will automatically mount cgroup2 fs.

oh and that is because the init container will mount the cgroup2 fs in that directory and then cilium container can use it?

joestringer replied:
Evidently past me was thinking about this more deeply than current me :sweat_smile: (Looking at the notes, I don't think we captured this nuance) . I think I recall the reasoning, it was something along the lines of, if the host isn't already mounting cgroupfs by itself then maybe /sys/fs/cgroup doesn't exist as a directory to mount to (because otherwise the OS would mount the FS there right?) and then maybe it's not our role to create that directory. Then it's better to just mount on `/var/run/cilium/cgroupv2` where we know that Cilium controls the directory and no-one should interfere with it.

OK, then I think this default is fine as-is.

aanm commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -92,6 +92,15 @@ configuration.
+.. note::
+
+    Cilium will automatically mount cgroup v2 filesystem required to attach BPF
+    cgroup programs by default at the path ``run/cilium/cgroupv2``. In order to
```
`run`->`/run`

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1180,6 +1181,7 @@ func initEnv(cmd *cobra.Command) {
 	// useful if the daemon is being round inside a namespace and the
 	// BPF filesystem is mapped into the slave namespace.
 	bpf.CheckOrMountFS(option.Config.BPFRoot, k8s.IsEnabled())
+	cgroups.CheckOrMountCgrpFS(option.Config.CGroupRoot)
```
any reason why can't this be on line 584 of daemon/cmd/daemon.go?

aditighag replied:
This is part of a revert commit.

aanm commented on pkg/cgroups/cgroups.go:
```diff
@@ -131,17 +56,6 @@ func CheckOrMountCgrpFS(mapRoot string, runsOnKind bool) {
 		// Failed cgroup2 mount is not a fatal error, sockmap will be disabled however
 		if err == nil {
```
Can you add a log message for the error condition? :pray:

aditighag replied:
This is fixed in the PR - https://github.com/cilium/cilium/pull/15999.

aditighag reviewed (commented):
@aanm I've addressed some of your comments. Replied inline to other comments. Thanks!

kkourt commented on Documentation/helm-values.rst:
```diff
@@ -90,6 +90,14 @@
+     - object
+     - ``{"hostRoot":"/run/cilium/cgroupv2"}``
+   * - cgroup.hostRoot
+     - Configure cgroup root where cgroup2 filesystem is mounted on the host (see also: `cgroup.autoMount`)
```
nit: if going patch-by-patch, the:

> (see also: `cgroup.autoMount`)

is added in a prior patch than the autoMount option.

Not blocking on my side.

kkourt reviewed (approved):
LGTM, thanks!

aditighag committed:
cilium-daemonset: Host cgroup root mount as alternative to auto-mount

Cilium agent daemonset auto mounts cgroup2 filesystem on the host by default.
However, it needs to mount host's `/proc` inside an init container
in order to do that. To disable this auto-mount behavior, we introduce a helm
option. When auto-mount is disabled, users can specify the mount point on the
underlying host where cgroup v2 fs is already mounted. We then volume mount this
directory inside the cilium agent pod.
The reason why we don't set the host cgroup2 mount point to a hard-coded path
such as `/sys/fs/cgroup`, is because cgroup2 filesystem mount point can be platform
dependent. See this note in the cgroup manpage [1] -

>Note that on many modern systems, systemd(1) automatically mounts
the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
process.

[1] https://man7.org/linux/man-pages/man7/cgroups.7.html

Suggested-by: Kornilios Kourtis <kornilios@isovalent.com>.
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aanm commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -92,6 +92,15 @@ configuration.
+    do that, it needs to mount the host ``/proc`` inside an init container
+    launched by the daemonset temporarily. If you need to disable the auto-mount,
+    specify ``--set cgroup.autoMount.enabled=false``, and set the host mount point
+    where cgroup v2 filesystem is already mounted by using ``--set cgroup.hostRoot``.
```
Can we have an example on how to mount cgroup2 for this case?

aanm reviewed (approved):
:tada:

joestringer commented on pkg/defaults/defaults.go:
```diff
@@ -98,7 +98,7 @@ const (
 
 	// DefaultCgroupRoot is the default path where cilium cgroup2 should be mounted
-	DefaultCgroupRoot = "/var/run/cilium/cgroupv2"
+	DefaultCgroupRoot = "/run/cilium/cgroupv2"
```
Any time that we change default options, we have the potential to impact upgrade, and the consequences are not always immediately clear. Not sure if there was any discussion on this particular change, but I've thought it through a bit below.

In practice this should not have any impact, unless there's the unusual combination of:
* Cilium is _not_ run in a container.
* The underlying distribution does _not_ link `/var/run` to `/run`.

Maybe in the above case, upgrading Cilium to this version with a new default could leave an extra mount of the cgroupfs behind which would never be unmounted. The user could of course unmount it.

This is all to say basically, I think the change is probably fine, but in future I'd suggest avoiding changes to defaults like this unless there's a clear reason for it and it's accompanied with upgrade impact analysis.

aditighag replied:
@joestringer FWIW, I did consider upgrade/downgrade implications before making the change. 

>upgrading Cilium to this version with a new default could leave an extra mount of the cgroupfs behind which would never be unmounted.

How would that happen? 🤔  `/var/run` is a symlink to `/run` on most platforms.

joestringer replied:
It's good to know that you had considered upgrade/downgrade implications. It can be a bit hard to gauge whether different developers are considering this on their PRs and whether we all have a shared understanding of the impacts. As someone trying to assess whether we should be including particular PRs in upcoming releases, my goal is basically to minimize the risk of the changes. If I see something that appears risky through some fuzzy logic (glancing at the changes, looking at what CI is saying) then I tend to look into what I can do to reduce that risk. The above comment was the output of that process with basically an assessment (low risk) and a suggestion (maybe in future we can avoid making changes to pkg/defaults).

One of the reasons I raised the comment above is that I see any change to `pkg/defaults` in a backport patch as a red flag. In general our upgrade model tends to assume that agent defaults do not change, with the notion in mind that users can end up using their existing DS YAMLs and just doing an image tag bump and they will not adopt new changes to existing settings. On the contrary if we want _new_ users to adopt some setting, then we tend to set the new value in the helm configuration and avoid making the change in the daemon. I'm not sure if we've ever formally encoded this knowledge somewhere or whether it's been effectively passed around the core team. (If we have passed this on successfully, then bear with me explaining it again :sweat_smile: )

The reason for this approach is that there's a certain class of upgrade bugs that can be caused by changing agent defaults. On one hand, banning changes to `pkg/defaults` in bugfix PRs isn't going to get rid of upgrade breakages. On the other hand, if it's possible to avoid changing default settings then I think we should avoid making those changes, so we don't have to consider this class of upgrade bugs.

joestringer commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -370,6 +375,28 @@ spec:
+          - sh
+          - -c
+          - 'mount | grep "$CGROUP_ROOT type cgroup2" || { echo "Mounting cgroup filesystem..."; mount -t cgroup2 none $CGROUP_ROOT; }'
+        image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}{{ if .Values.image.useDigest }}@{{ .Values.image.digest }}{{ end }}"
```
@aditighag btw just while looking back over this implementation I noticed that this container is missing `imagePullPolicy`. We shuold probably set it the same way as the other init container below.
